### PR TITLE
WIP: Register JITTed functions with libunwind

### PR DESCRIPTION
**Pull request by Keno**
I am extremely surprised that we didn't do this before - I guess libunwind
was just pretty good at figuring it out anyway, unlike the OS X unwind
library. However, for better backtraces, register the JITted functions
with libunwind, which also prevents libunwind from trying to access
invalid memory in a good number of cases.

WIP, because libunwind only supports 32 bit offsets between the .text
and .eh_frame sections, but MCJIT spreads the farther apart. I am in
the process of fixing this in libunwind.

### CODE REVIEW

File: src/debuginfo.cpp [hunk 1]
```diff
@@ -952,6 +958,119 @@ RTDyldMemoryManager* createRTDyldMemoryManagerOSX() {
...
+};
+
+static uint8_t *consume_leb128(uint8_t *Addr, size_t Size) {
+    uint8_t *P = Addr;
+    while ((*P >> 7) != 0 && P < Addr + Size)
+        ++P;
+    return P;
+}
+
+#include <iostream>
```
Comment (ihnorton) [inline src/debuginfo.cpp]: :left_luggage:

Comment (Keno) [inline src/debuginfo.cpp]: thanks

File: src/debuginfo.cpp [hunk 2]
```diff
@@ -865,6 +869,31 @@ int jl_get_llvmf_info(uint64_t fptr, uint64_t *symsize, uint64_t *slide,
 #endif
 }
 
+#if defined(LLVM37) && (defined(_OS_LINUX_) || (defined(_OS_DARWIN_) && defined(LLVM_SHLIB)))
+extern "C" void __register_frame(void*);
+extern "C" void __deregister_frame(void*);
+
+template <typename callback>
+static const char *processFDE(const char *Entry, callback f) {
+  const char *P = Entry;
```
Comment (tkelman) [inline src/debuginfo.cpp]: should be 4-space indent here and below, at least for the code this PR touches?

**Comment by Keno** [comment]
I just mailed an appropriate patch upstream. Will update here once it's accepted.

**Comment by yuyichao** [comment]
Do you have a reference for it?

Also, do you have any idea if this could be the cause of https://github.com/JuliaLang/julia/issues/11691 ? I tried to enable the test on top of your branch and it doesn't fixed the problem but maybe it could with the libunwind patch.

**Comment by Keno** [comment]
Yes, quite possibly. The failure mode is is a ReadOnlyMemoryError. However, note that this is LLVM 3.7 only, because the old JIT did not emit the appropriate unwind info.

**Comment by JeffBezanson** [comment]
Yes, could you post a gist of the patch just in case?

**Comment by Keno** [comment]
https://gist.github.com/Keno/152e0eb513973dad5726

**Comment by carnaval** [comment]
@Keno any news ? libunwind often stops unwinding early with llvm-svn on jited code. I assume that it's because we often emit functions without frame pointers now, although I'm not sure what libunwind fallbacks to when proper unwind info is not provided.

**Comment by carnaval** [comment]
also if it's less bothersome than libunwind upstream we can probably fix that by overriding the default allocator in the SectionManager thing.

**Comment by Keno** [comment]
I reminded upstream about the patch. Let's see if it sticks.

**Comment by yuyichao** [comment]
Seems that the patc have been committed a few month ago? http://git.savannah.gnu.org/gitweb/?p=libunwind.git;a=commit;h=8afc33ce9f20538f048838e18680c120867ee0d2

**Comment by Keno** [comment]
Oh, huh. That's nice! Nobody told me about it.

**Comment by yuyichao** [comment]
There also seems to be other improvements on libunwind master since I'm getting much better backtrace with it even without yours or @carnaval 's patch. I'm wondering when will they make a new release (given that the last one is more than 3yrs ago).

**Comment by tkelman** [comment]
Hopefully soon after they merge the musl patch I sent them.

**Comment by carnaval** [comment]
Alright, what is holding this up ? Can we move on to libunwind master/carry the patch in our tree and merge this ? @Keno ?

We need this also to get inlined frame in backtraces if we start using DWARF for that since even though libunwind guesses correctly it certainly can't invent those ghost frames.

**Comment by yuyichao** [comment]
I just rebased this on current master (old branch backed up at [yyc/kf/linuxunwind](https://github.com/JuliaLang/julia/tree/yyc/kf/linuxunwind)). Compilation passes with libunwind master but given this needs the `UNW_INFO_FORMAT_IP_OFFSET` format which is not available on the released version of libunwind, the linux CI will probably fail.

The libunwind master has a lot of features we need (much better unwind on x86 (and no segfault), aarch64 support, new dynamic unwind info format) and since there hasn't been a release since 3 years ago, Given libunwind isn't moving as fast as LLVM, I think we could try to ship the git master version of libunwind (fetch from git/make our own tarball) instead of the current released one. (Unless someone can convince libunwind to make a new release soon)

**Comment by yuyichao** [comment]
I pushed a commit to use my own tarball built from the libunwind master, let's see what the CI thinks.

**Comment by tkelman** [comment]
Can we move that somewhere more permanent? Or bother upstream into making a new release?

**Pull request merged**